### PR TITLE
feat(prettier): add id on `IfBreak` and `Group`

### PR DESCRIPTION
**Pull request by mysteryven**

### CODE REVIEW

File: crates/oxc_prettier/src/doc.rs [hunk 1]
```diff
@@ -37,22 +40,36 @@ pub enum Doc<'a> {
...
 }
 
+static ID_COUNTER: AtomicUsize = AtomicUsize::new(0);
+pub type DocId = u32;
+
+impl<'a> Doc<'a> {
+    pub fn next_id() -> DocId {
+        DocId::try_from(ID_COUNTER.fetch_add(1, Ordering::SeqCst)).unwrap_or_default()
+    }
+}
```
[the remark this answers is not in the document]
Comment (Boshen) [reply crates/oxc_prettier/src/doc.rs:57]: I agree, we don't need atomics for tracking state.

[the remark this answers is not in the document]
Comment (mysteryven) [reply crates/oxc_prettier/src/doc.rs]: Done, only use a `u32` without inner mutability, it's not needed current.

File: crates/oxc_prettier/src/lib.rs [hunk 2]
```diff
@@ -23,6 +23,19 @@ use crate::{doc::Doc, format::Format, printer::Printer};
 
 pub use crate::options::{ArrowParens, EndOfLine, PrettierOptions, QuoteProps, TrailingComma};
 
+type GroupId = u32;
+#[derive(Default)]
+struct GroupIdBuilder {
```
Comment (Boshen) [inline crates/oxc_prettier/src/lib.rs:28]: Let's use the newtype idiom here https://doc.rust-lang.org/rust-by-example/generics/new_types.html

Just call it `GroupId` and expose the API `next_id`.

File: crates/oxc_prettier/src/printer/mod.rs [hunk 3]
```diff
@@ -220,9 +228,18 @@ impl<'a> Printer<'a> {
         self.line_suffix.push(Command { indent, mode, doc: Doc::Array(docs) });
     }
 
-    fn handle_if_break(&mut self, doc: Doc<'a>, indent: Indent, mode: Mode) {
-        if mode == Mode::Break {
-            self.cmds.push(Command::new(indent, Mode::Break, doc));
+    fn handle_if_break(&mut self, if_break: IfBreak<'a>, indent: Indent, mode: Mode) {
+        let IfBreak { break_contents, flat_content, group_id } = if_break;
+        let group_mode = group_id.map_or(Some(mode), |id| self.group_mode_map.get(&id).copied());
```
Comment (camc314) [inline crates/oxc_prettier/src/printer/mod.rs:233]:
```suggestion
        let group_mode = group_id.map_or(mode, |id| self.group_mode_map.get(&id).cloned().unwrap_or(Mode::Flat));
```
i don't think this should return an option - we should always print either the contents of `flat` or `break` regardless of whether the groupid was found in the map.

prehaps we should `dbg_assert!` that it exists?

Comment (camc314) [reply crates/oxc_prettier/src/printer/mod.rs:233]: prettier for ref -> https://github.com/prettier/prettier/blob/eebf0e4b5ec8ac24393c56ced4b4819d4c551f31/src/document/printer.js#L269-L271

Comment (camc314) [reply crates/oxc_prettier/src/printer/mod.rs:233]: oh my bad, was looking at `fits`

File: crates/oxc_prettier/src/lib.rs (hunk 2 above)
Comment (mysteryven) [reply crates/oxc_prettier/src/lib.rs]: Done!

**Comment by Boshen** [comment]
### Merge activity

* **Nov 27, 9:53 AM**: @Boshen merged this pull request with [Graphite](https://app.graphite.dev/github/pr/oxc-project/oxc/1551).

**Pull request merged**